### PR TITLE
Convert EntityTile, MemberTile and PresenceLabel to TS

### PR DESCRIPTION
**Pull request by SimonBrandner**

### CODE REVIEW

File: src/components/views/rooms/EntityTile.tsx [hunk 1]
```diff
@@ -16,29 +16,34 @@ See the License for the specific language governing permissions and
...
+}
 
 const PRESENCE_CLASS = {
     "offline": "mx_EntityTile_offline",
     "online": "mx_EntityTile_online",
     "unavailable": "mx_EntityTile_unavailable",
 };
 
-function presenceClassForMember(presenceState, lastActiveAgo, showPresence) {
+function presenceClassForMember(presenceState: string, lastActiveAgo: number, showPresence: boolean) {
```
Comment (t3chguy) [inline src/components/views/rooms/EntityTile.tsx:39]: return type please

File: src/components/views/rooms/EntityTile.tsx [hunk 2]
```diff
@@ -51,42 +56,52 @@ function presenceClassForMember(presenceState, lastActiveAgo, showPresence) {
...
+    presenceLastTs?: number,
+    presenceCurrentlyActive?: boolean,
+    showInviteButton?: boolean,
+    shouldComponentUpdate?(nextProps: IProps, nextState: IState): boolean,
+    onClick?(): void,
+    suppressOnHover?: boolean,
+    showPresence?: boolean,
+    subtextLabel?: string,
+    e2eStatus?: string,
+    powerStatus?: string,
```
Comment (t3chguy) [inline src/components/views/rooms/EntityTile.tsx:75]: can this be enummed?

File: src/components/views/rooms/EntityTile.tsx [hunk 3]
```diff
@@ -16,29 +16,34 @@ See the License for the specific language governing permissions and
...
-import { _t } from '../../../languageHandler';
+import { _td } from '../../../languageHandler';
 import classNames from "classnames";
 import E2EIcon from './E2EIcon';
-import {replaceableComponent} from "../../../utils/replaceableComponent";
+import { replaceableComponent } from "../../../utils/replaceableComponent";
+import BaseAvatar from '../avatars/BaseAvatar';
+import PresenceLabel from "./PresenceLabel";
+
+const PowerLabel: Record<string, string> = {
```
Comment (t3chguy) [inline src/components/views/rooms/EntityTile.tsx:28]: enum key please

File: src/components/views/rooms/EntityTile.tsx [hunk 4]
```diff
@@ -51,42 +56,52 @@ function presenceClassForMember(presenceState, lastActiveAgo, showPresence) {
...
+    name?: string,
+    title?: string,
+    avatarJsx?: JSX.Element, // <BaseAvatar />
+    className?: string,
+    presenceState?: string,
+    presenceLastActiveAgo?: number,
+    presenceLastTs?: number,
+    presenceCurrentlyActive?: boolean,
+    showInviteButton?: boolean,
+    shouldComponentUpdate?(nextProps: IProps, nextState: IState): boolean,
```
Comment (t3chguy) [inline src/components/views/rooms/EntityTile.tsx:69]: this prop looks to never get set, can we please kill it with fire

File: src/components/views/rooms/EntityTile.tsx (hunk 3 above)
Comment (t3chguy) [reply src/components/views/rooms/EntityTile.tsx:28]: They were enum-like before

File: src/components/views/rooms/EntityTile.tsx [hunk 5]
```diff
@@ -168,14 +179,12 @@ class EntityTile extends React.Component {
...
 
-        const BaseAvatar = sdk.getComponent('avatars.BaseAvatar');
-
         const av = this.props.avatarJsx ||
             <BaseAvatar name={this.props.name} width={36} height={36} aria-hidden="true" />;
 
         // The wrapping div is required to make the magic mouse listener work, for some reason.
         return (
-            <div ref={(c) => this.container = c} >
+            <div ref={this.container} >
```
Comment (t3chguy) [inline src/components/views/rooms/EntityTile.tsx:187]: I can't see anything actually using this ref, especially as its signature is changing from a LegacyRef to a RefObject

Comment (t3chguy) [reply src/components/views/rooms/EntityTile.tsx:187]:
```suggestion
            <div ref={this.container}>
```

File: src/components/views/rooms/MemberTile.tsx [hunk 6]
```diff
@@ -17,20 +17,32 @@ limitations under the License.
...
-import {replaceableComponent} from "../../../utils/replaceableComponent";
+import { replaceableComponent } from "../../../utils/replaceableComponent";
+import { RoomMember } from "matrix-js-sdk/src/models/room-member";
+import { MatrixEvent } from "matrix-js-sdk/src/models/event";
+import { EventType } from "matrix-js-sdk/src/@types/event";
+import { DeviceInfo } from "matrix-js-sdk/src/crypto";
+
+interface IProps {
+    member: RoomMember,
+    showPresence?: boolean,
```
Comment (t3chguy) [inline src/components/views/rooms/MemberTile.tsx:33]:
```suggestion
    member: RoomMember;
    showPresence?: boolean;
```

File: src/components/views/rooms/EntityTile.tsx [hunk 7]
```diff
@@ -51,42 +56,52 @@ function presenceClassForMember(presenceState, lastActiveAgo, showPresence) {
...
-        presenceCurrentlyActive: PropTypes.bool,
-        showInviteButton: PropTypes.bool,
-        shouldComponentUpdate: PropTypes.func,
-        onClick: PropTypes.func,
-        suppressOnHover: PropTypes.bool,
-        showPresence: PropTypes.bool,
-        subtextLabel: PropTypes.string,
-        e2eStatus: PropTypes.string,
-    };
+interface IProps {
```
Comment (t3chguy) [inline src/components/views/rooms/EntityTile.tsx:59]: s/,/;/g please

File: src/components/views/rooms/MemberTile.tsx [hunk 8]
```diff
@@ -17,20 +17,32 @@ limitations under the License.
...
+}
+
+interface IState {
+    statusMessage: string,
+    isRoomEncrypted: boolean,
+    e2eStatus: string,
```
Comment (t3chguy) [inline src/components/views/rooms/MemberTile.tsx:39]:
```suggestion
    statusMessage: string;
    isRoomEncrypted: boolean;
    e2eStatus: string;
```

File: src/components/views/rooms/MemberTile.tsx [hunk 9]
```diff
@@ -143,32 +155,32 @@ export default class MemberTile extends React.Component {
...
     }
 
-    getStatusMessage() {
+    private getStatusMessage(): string {
         const { user } = this.props.member;
         if (!user) {
             return "";
         }
-        return user._unstable_statusMessage;
+        return user.unstable_statusMessage;
```
Comment (t3chguy) [inline src/components/views/rooms/MemberTile.tsx:163]: whoops

File: src/components/views/rooms/PresenceLabel.tsx [hunk 10]
```diff
@@ -15,56 +15,53 @@ limitations under the License.
...
-    };
+interface IProps {
+    // number of milliseconds ago this user was last active.
+    // zero = unknown
+    activeAgo?: number,
+    // if true, activeAgo is an approximation and "Now" should
+    // be shown instead
+    currentlyActive?: boolean,
+    // offline, online, etc
+    presenceState?: string,
```
Comment (t3chguy) [inline src/components/views/rooms/PresenceLabel.tsx:30]:
```suggestion
    activeAgo?: number;
    // if true, activeAgo is an approximation and "Now" should
    // be shown instead
    currentlyActive?: boolean;
    // offline, online, etc
    presenceState?: string;
```

File: src/components/views/rooms/PresenceLabel.tsx [hunk 11]
```diff
@@ -15,56 +15,53 @@ limitations under the License.
 */
 
 import React from 'react';
-import PropTypes from 'prop-types';
 
 import { _t } from '../../../languageHandler';
 import {replaceableComponent} from "../../../utils/replaceableComponent";
```
Comment (t3chguy) [inline src/components/views/rooms/PresenceLabel.tsx:20]:
```suggestion
import { replaceableComponent } from "../../../utils/replaceableComponent";
```

**Review by t3chguy — COMMENTED** [review]
looks good otherwise

File: src/components/views/rooms/EntityTile.tsx (hunk 4 above)
Comment (SimonBrandner) [reply src/components/views/rooms/EntityTile.tsx:69]: :fire:

File: src/components/views/rooms/EntityTile.tsx (hunk 5 above)
Comment (SimonBrandner) [reply src/components/views/rooms/EntityTile.tsx:187]: Yep, it doesn't seem used

File: src/components/views/rooms/MemberTile.tsx (hunk 9 above)
Comment (SimonBrandner) [reply src/components/views/rooms/MemberTile.tsx:163]: [This seems alright](https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/models/user.ts#L46)?

Comment (t3chguy) [reply src/components/views/rooms/MemberTile.tsx:163]: Yup, whoops that I left it broken and for you to fix :P

Comment (SimonBrandner) [reply src/components/views/rooms/MemberTile.tsx:163]: Ah! :smile:

File: src/components/views/rooms/EntityTile.tsx [hunk 12]
```diff
@@ -17,28 +17,38 @@ limitations under the License.
...
+import PresenceLabel from "./PresenceLabel";
+
+enum PowerStatus {
+    Admin = "admin",
+    Moderator = "moderator",
+}
+
+const PowerLabel: Record<PowerStatus, string> = {
+    "admin": _td("Admin"),
+    "moderator": _td("Mod"),
```
Comment (t3chguy) [inline src/components/views/rooms/EntityTile.tsx:35]:
```suggestion
    [PowerStatus.Admin]: _td("Admin"),
    [PowerStatus.Moderator]: _td("Mod"),
```

File: src/components/views/rooms/EntityTile.tsx [hunk 13]
```diff
@@ -82,13 +95,16 @@ class EntityTile extends React.Component {
...
+        this.state = {
+            hover: false,
+        };
+    }
 
-    shouldComponentUpdate(nextProps, nextState) {
+    shouldComponentUpdate(nextProps: IProps, nextState: IState) {
         if (this.state.hover !== nextState.hover) return true;
-        return this.props.shouldComponentUpdate(nextProps, nextState);
     }
```
Comment (t3chguy) [inline src/components/views/rooms/EntityTile.tsx:108]: `shouldComponentUpdate` defaultProp used to be `() => true` so we either need a `return true` here but I think we can just forgo the shouldComponentUpdate and switch this to a PureComponent

**Review by t3chguy — CHANGES_REQUESTED** [review]

**Comment by t3chguy** [comment]
Blocked by request of @SimonBrandner

**Pull request merged**